### DISCUSSION TITLE
Thread-safe method to transport the `request` object over a request in Rails

### DISCUSSION THREAD

**Discussion by adrianthedev** [ioquatix/ioquatix#17] [Q&A]
Hey there 👋

I'm Adrian, the author of Avo, and I'm in a bit of a pickle with one design decision I took in the beginning.
I was nudged here by Joel Drapper. He said "Samuel Williams is the concurrency expert".

I have a concurrency issue that I don't yet know how to fix in a nice way. At least, I think it's a concurrency issue. I tested it by reducing puma's `max_threads_count` to `1`, and the issue does not persist.

### Intro

I'm doing two things at request-time:

1. I do some initialization by finding some classes, instantiating them to objects, and holding them in an array so I can request them throughout the app when I need them.
2. I need to keep the `request` object (among others) somewhere for the same reason, to request it later throughout the app.

I'm doing that using a few `class_attribute`s on a class (pseudo-code below).

```ruby
# lib/app.rb
class App
  class_attribute :resource, default: []
  class_attribute :request

  def self.init(request:,)
    self.request = request
    self.resources = fetch_resources
  end
end

# engine's application_controller.rb
def ApplicationController
  before_action :init_app

  def init_app
    App.init request: request
  end
end

# Somewhere deep in the app I use them like so.
do_something_with_the_params App.request.params
do_something_with_resources App.request.resources
```

### The issue

The bug is that if the app gets a lot of requests at once, there are "weird" things happening. Some objects are not available in memory and the page crashes. When I refresh that page, the issue resolves itself.

I can't precisely put my finger on what's happening, but it's definitely related to that design choice.
Also, I'm aware that that's bad design; it just evolved like that, but I'm trying to fix it.

### The question

How can I safely (with a nice, maintainable API) hold those objects in memory on a per-request basis?

Thanks for your help 🙏

**Comment by rajaravivarma-r** [comment]
Hi @adrianthedev, We used to use `Thread.current[:attribute] = value` with great success.

Pseudocode:
`app/models/user.rb`
```ruby
class User
  def self.current=(user)
    Thread.current[:current_user] = user
  end
  
  def self.current
    Thread.current[:current_user]
  end
end
```

`app/controllers/application_controller.rb`
```ruby
class ApplicationController
  around_action: set_reset_current_user
  
  private
  
  def set_reset_current_user
    User.current = user
  ensure
    User.current = nil # or User.reset_current_user
  end
end
```

[RequestStore ](https://github.com/steveklabnik/request_store) Automates it using a Middleware and it works great.

More recently, I have come upon [ActiveSupport::CurrentAttributes](https://api.rubyonrails.org/classes/ActiveSupport/CurrentAttributes.html), but I haven't used it.

**Comment by ioquatix** [comment]
I am working on a proper solution to this problem, and it is a problem, and you aren't the first one to have it.

https://github.com/ruby/ruby/pull/6612

The idea is that writing `Fiber[key] = value` should propagate to child fibers and threads spawned from within the current fiber scope.

This means, with the above PR, you'd just write `Fiber[:current_user] = ...` and `Fiber[:current_user]` would work in all nested contexts.

This should provide a good answer to "How can I safely (with a nice, maintainable API) hold those objects in memory on a per-request basis?" at the language level.

**Reply by adrianthedev** [reply]
That sounds good! I love that these kinds of matters have someone looking into them and fixing them forever.

Until that feature launches, would you say that one could reliably use `Thread.current[:current_user]` as a temporary solution? What downsides do you forsee? 
*I'm happy to read an article or something someone else published to get up to speed with this matter.

**Reply by ioquatix** [reply]
Unfortunately, as you have found out, `Thread.current` is the best we have, but still a very poor substitute.

As you might already know, it breaks with `Enumerator` and other code which doesn't correctly handle thread and fiber locals correctly, hence my new PR to try and provide an interface for what most users actually want.
